cmd/service-watcher/main.go

```diff
@@ -0,0 +1,178 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	systemdbus "github.com/coreos/go-systemd/dbus"
+
+	"github.com/TheCacophonyProject/event-reporter/eventstore"
+)
+
+const (
+	minTimeBetweenReports = 20 * time.Minute //TODO add into cacophony-config
+	logLines              = 20               //TODO add into cacophony-config
+)
+
+type LogReport struct {
+	Logs        *[]string
+	Time        int64
+	UnitName    string
+	ActiveState string
+}
+
+type LogsRaw struct {
+	Unit              string `json:"UNIT"`
+	SystemdUnit       string `json:"_SYSTEMD_UNIT"`
+	RealtimeTimestamp string `json:"__REALTIME_TIMESTAMP"`
+	Message           string `json:"MESSAGE"`
+}
+
+func main() {
+	err := runMain()
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+func runMain() error {
+	log.SetFlags(0) // Removes default timestamp flag
+
+	conn, err := systemdbus.New()
+	if err != nil {
+		log.Println("failed to connect to dbus")
+		return err
+	}
+
+	defer conn.Close()
+
+	if err := conn.Subscribe(); err != nil {
+		log.Println("failed to subscribe to the dbus")
+		return err
+	}
+
+	updateCh := make(chan *systemdbus.PropertiesUpdate, 256)
+	errCh := make(chan error, 256)
+	conn.SetPropertiesSubscriber(updateCh, errCh)
+
+	lastUnitReportTimes := map[string]time.Time{}
+
+	for {
+		select {
+		case update := <-updateCh:
+			ts := time.Now()
+			activeState := strings.Trim(update.Changed["ActiveState"].String(), "\"")
+			unitName := update.UnitName
+			// Only process states we are interested in
+			if !inActiveStates(activeState) {
+				break
+			}
+			if t, ok := lastUnitReportTimes[unitName]; ok && time.Now().Sub(t) < minTimeBetweenReports {
+				log.Println("reporting too often")
+				break
+			}
+
+			log.Println("unitname:", unitName)
+			log.Println("activeState:", activeState)
+
+			rawLogs, failed, err := getLogs(update.UnitName, logLines)
+			if err != nil {
+				return err
+			}
+			if !failed {
+				break // Can just be a service activating
+			}
+
+			event := eventstore.Event{
+				Timestamp: ts,
+				Description: eventstore.EventDescription{
+					Type: "systemError",
+					Details: map[string]interface{}{
+						"versoin":     1,
+						"unitName":    unitName,
+						"logs":        rawLogs,
+						"activeState": activeState,
+					},
+				},
+			}
+			if err != nil {
+				return err
+			}
+			if err := eventstore.AddEvent(event); err != nil {
+				return err
+			}
+			lastUnitReportTimes[unitName] = time.Now()
+
+		case err := <-errCh:
+			log.Println("err:", err)
+		}
+	}
+}
+
+func getLogs(unitName string, number int) (*[]string, bool, error) {
```

Comment:
There's no need to return the lines as a pointer to a slice. A slice alone will do.